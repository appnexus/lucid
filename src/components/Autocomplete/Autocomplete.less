@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

@Autocomplete-size-menu-maxHeight: 220px;
@Autocomplete-size-controlHeight: @size-standard-height;

.lucid-Autocomplete {

	&:not(.lucid-ContextMenu-FlyOut) {
		width: 100%;
	}

	.lucid-ContextMenu {
		width: 100%;
	}

	& &-Control {
		.no-safari-flicker();
		.box-sizing();
		cursor: pointer;
		height: @Autocomplete-size-controlHeight;
		display: flex;
		outline: none;
		border: 1px solid @color-borderColor;
		border-radius: @size-borderRadius;
		font: @font;
		color: @color-black;
		letter-spacing: .05em;
		line-height: @size-standard-height;
		padding: 0 @size-S;
		align-items: center;

		// The highly specific selector makes it so global input styles
		// are less likely to affect the Autocomplete input.
		& .lucid-Autocomplete-Control-input {
			background: none @color-transparent;
			border: 0 solid @color-transparent;
			outline: none;
			font: @font;
			font-weight: normal;
			color: @color-textColor;
			margin: 0;
			padding: 0;
			width: 100%;
		}

		.lucid-CaretIcon {
			fill: @color-black;
			width: 8px;
			height: 8px;
			margin-left: auto;
		}

		// States

		&-is-expanded {
			border-color: @color-primary;
		}

		&-is-disabled {
			cursor: not-allowed;
			.opacity();
		}

		&-is-disabled {
			input {
				cursor: not-allowed;
			}
		}
	}

	&.lucid-DropMenu-is-expanded {
		&.lucid-DropMenu-direction-up {
			.lucid-Autocomplete-Control {
				border-radius: 0 0 @size-borderRadius @size-borderRadius;
			}
		}

		&.lucid-DropMenu-direction-down {
			.lucid-Autocomplete-Control {
				border-radius: @size-borderRadius @size-borderRadius 0 0;
			}
		}

		.lucid-DropMenu-Control {
			z-index: @zindex-tooltip;
		}
	}

	&.lucid-ContextMenu-FlyOut {
		max-height: @Autocomplete-size-menu-maxHeight;
	}

	&-Option-suggestion-match,
	&-Option-suggestion-post,
	&-Option-suggestion-pre {
		white-space: pre;
	}

	&-Option-suggestion-match {
		text-decoration: underline;
	}
}
