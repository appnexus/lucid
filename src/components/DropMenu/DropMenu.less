@import (reference) '../../styles/variables.less';
@import (reference) '../../styles/mixins.less';

@DropMenu-size-optionHeight: 25px;

.lucid-DropMenu {

	&&-base {
		display: inline-block;
	}

	&-Header {
		border-top: solid 1px @featured-color-default-borderColor;
		border-left: solid 1px @featured-color-default-borderColor;
		border-right: solid 1px @featured-color-default-borderColor;
	}

	&-Control {
		position:relative;
		outline: 0px;
	}

	&-Option {
		cursor: pointer;
		padding: 0 @size-S;
		line-height: @size-standard;
		font: @font;
		height: @DropMenu-size-optionHeight;
		display: flex;
		align-items: center;
		white-space: nowrap;

		&&-is-disabled {
			color: @color-disabledText;
			font-style: italic;
			cursor: not-allowed;
		}

		&&-is-focused {
			background: @color-lightGray;
		}

		&&-is-selected {
			background: @color-primaryMedium;
		}

		&&-is-null {
			color: @color-disabledText;
		}

		&&-is-wrapped {
			white-space: normal;
		}
	}

	&-option-container {
		background-color: @color-white;
		border: solid 1px @featured-color-default-borderColor;
		.box-sizing();
		overflow-y: auto;

		// TODO: do we stil need this?
		// @supports (-webkit-appearance:none)
		// 	and (text-align-last:auto) and (not (-ms-accelerator:true)) {
		// 		-webkit-transform:translate3d(0,0,0); /* makes the scrollbar visible */
		// }
	}

	&.lucid-ContextMenu-FlyOut {
		&&-Up {
			border-width: 1px 1px 0 1px;
			border-radius: @size-borderRadius @size-borderRadius 0 0;
			box-shadow: 0 1px 0 @featured-color-default-borderColor;
		}

		&&-Down {
			border-width: 0 1px 1px 1px;
			border-radius: 0 0 @size-borderRadius @size-borderRadius;
			box-shadow: 0 -1px 0 @featured-color-default-borderColor;
		}
	}

	&-OptionGroup-divider {
		border: 1px solid @color-gray;
		border-width: 1px 0 0 0;
		height: 1px;
	}

	&-label {
		padding: 0 @size-XS;
		font: @font;
		font-weight: bold;
		height: @DropMenu-size-optionHeight;
		display: flex;
		align-items: center;
	}
}
