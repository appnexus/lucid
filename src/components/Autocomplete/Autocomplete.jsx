import React from 'react';
import _ from 'lodash';
import { lucidClassNames } from '../../util/style-helpers';
import { createClass } from '../../util/component-definition';
import * as reducers from './Autocomplete.reducers';
import * as KEYCODE from '../../constants/key-code';
import DropMenu from '../DropMenu/DropMenu';
import CaretIcon from '../Icon/CaretIcon/CaretIcon';

const boundClassNames = lucidClassNames.bind('&-Autocomplete');

const {
	arrayOf,
	bool,
	func,
	object,
	shape,
	string
} = React.PropTypes;

/**
 *
 * {"categories": ["controls", "text"], "madeFrom": ["CaretIcon", "DropMenu"]}
 *
 * A text input with suggested values displayed in an attached menu.
 */

const Autocomplete = createClass({
	displayName: 'Autocomplete',

	reducers,

	propTypes: {
		/**
		 * Appended to the component-specific class names set on the root elements.
		 */
		className: string,
		/**
		 * Styles that are passed through to root element.
		 */
		style: object,
		/**
		 * Disables the Autocomplete from being clicked or focused.
		 */
		isDisabled: bool,
		/**
		 * Array of suggested text input values shown in drop menu.
		 */
		suggestions: arrayOf(string),
		/**
		 * Text value of the input.
		 */
		value: string,
		/**
		 * Object of DropMenu props which are passed thru to the underlying DropMenu component.
		 */
		DropMenu: shape(DropMenu.propTypes),
		/**
		 * Called when the input value changes.
		 * Has the signature `(value, {props, event}) => {}` where value is a string.
		 */
		onChange: func,
		/**
		 * Called when a suggestion is selected from the menu.
		 * Has the signature `(optionIndex, {props, event}) => {}` where optionIndex is a number.
		 */
		onSelect: func,
		/**
		 * Called when menu is expected to expand.
		 * Has the signature `({props, event}) => {}`.
		 */
		onExpand: func,
	},

	getDefaultProps() {
		return {
			isDisabled: false,
			suggestions: [],
			value: '',
			onChange: _.noop,
			onSelect: _.noop,
			onExpand: _.noop,
			DropMenu: DropMenu.getDefaultProps()
		};
	},

	handleSelect(optionIndex, { event }) {
		const {
			suggestions,
			onChange,
			onSelect,
		} = this.props;

		onChange(suggestions[optionIndex], { event, props: this.props});
		onSelect(optionIndex, { event, props: this.props});
	},

	handleInput(event) {
		const {
			onChange,
			onExpand,
			DropMenu: {
				onCollapse
			}
		} = this.props;

		onChange(event.target.value, {event, props: this.props});
		if (!_.isEmpty(event.target.value)) {
			onExpand({event, props: this.props});
		} else {
			onCollapse();
		}
	},

	getInputValue() {
		return _.get(this.refs, 'inputNode.value', this.props.value);
	},

	setInputValue(value) {
		if (this.refs.inputNode) {
			this.refs.inputNode.value = value;
		}
	},

	handleInputKeydown(event) {
		const {
			onExpand,
			DropMenu: {
				isExpanded,
				focusedIndex,
				onCollapse,
			}
		} = this.props;

		const value = this.getInputValue();

		if (event.keyCode === KEYCODE.Tab && isExpanded && focusedIndex !== null) {
			this.handleSelect(focusedIndex, {event, props: this.props});
			event.preventDefault();
		}

		if (event.keyCode === KEYCODE.Space) {
			event.stopPropagation();
		}

		if (event.keyCode === KEYCODE.ArrowDown && !isExpanded) {
			event.stopPropagation();

			if (_.isEmpty(value)) {
				onExpand({event, props: this.props});
			}
		}

		if (event.keyCode === KEYCODE.Escape) {
			event.stopPropagation();
			onCollapse(event);
		}

		if (event.keyCode === KEYCODE.Enter && focusedIndex === null) {
			event.stopPropagation();
			onCollapse(event);
		}
	},

	handleControlClick(event) {
		const {
			onExpand,
			DropMenu: {
				isExpanded,
				onCollapse
			}
		} = this.props;

		if (event.target === this.refs.inputNode) {
			onExpand({event, props: this.props});
		} else {
			if (isExpanded) {
				onCollapse(event);
			} else {
				onExpand({event, props: this.props});
			}

			this.refs.inputNode.focus();
		}
	},

	componentDidMount() {
		const { value } = this.props;
		this.refs.inputNode.addEventListener('input', this.handleInput);
		this.setInputValue(value);
	},

	componentWillReceiveProps(nextProps) {
		const { value } = nextProps;
		if (value !== this.getInputValue()) {
			this.setInputValue(value);
		}
	},

	componentWillUnmount() {
		if (this.refs.inputNode) {
			this.refs.inputNode.removeEventListener('input', this.handleInput);
		}
	},

	render() {
		const {
			style,
			className,
			isDisabled,
			DropMenu: dropMenuProps,
			suggestions,
			...passThroughs
		} = this.props;

		const {
			direction,
			isExpanded
		} = dropMenuProps;

		const value = this.getInputValue();
		const valuePattern = new RegExp(_.escapeRegExp(value), 'i');

		return (
			<DropMenu
				{...dropMenuProps}
				isDisabled={isDisabled}
				selectedIndices={[]}
				className={boundClassNames('&', className)}
				onSelect={this.handleSelect}
				style={style}
			>
				<DropMenu.Control onClick={this.handleControlClick}>
					<div className={boundClassNames('&-Control', {
						'&-Control-is-expanded': isExpanded,
						'&-Control-is-disabled': isDisabled
					})}>
						<input
							{..._.omit(passThroughs, ['onChange', 'onSelect', 'onExpand', 'value', 'children'])}
							type='text'
							className={boundClassNames('&-Control-input')}
							ref='inputNode'
							onKeyDown={this.handleInputKeydown}
							disabled={isDisabled}
						/>
						<CaretIcon direction={isExpanded ? direction : 'down'} />
					</div>
				</DropMenu.Control>
				{value ? _.map(suggestions, (suggestion) => (
					<DropMenu.Option key={'AutocompleteOption' + suggestion} >
						{(() => {
							const [pre, match, post] = partitionText(suggestion, valuePattern, value.length);
							const formattedSuggestion = [];
							if (pre) {
								formattedSuggestion.push(
									<span key={`AutocompleteOption-suggestion-pre-${suggestion}`} className={boundClassNames('&-Option-suggestion-pre')}>{pre}</span>
								);
							}
							if (match) {
								formattedSuggestion.push(
									<span key={`AutocompleteOption-suggestion-match-${suggestion}`} className={boundClassNames('&-Option-suggestion-match')}>{match}</span>
								);
							}
							if (post) {
								formattedSuggestion.push(
									<span key={`AutocompleteOption-suggestion-post-${suggestion}`} className={boundClassNames('&-Option-suggestion-post')}>{post}</span>
								);
							}
							return formattedSuggestion;
						})()}
					</DropMenu.Option>)
				) : _.map(suggestions, (suggestion) => (
					<DropMenu.Option key={'AutocompleteOption' + suggestion} >{suggestion}</DropMenu.Option>)
				)}
			</DropMenu>
		);
	}
});

function partitionText(text, pattern, length) {
	const index = text.search(pattern);
	if (index === -1) {
		return ['', '', text];
	} else if (index === 0) {
		return ['', text.substr(0, length), text.substring(length)];
	} else {
		return [text.substring(0, index), text.substr(index, length), text.substring(index + length)];
	}
}

export default Autocomplete;
