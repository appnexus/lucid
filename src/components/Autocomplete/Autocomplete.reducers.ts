import _ from 'lodash';
import * as DropMenu from '../DropMenu/DropMenu.reducers';

export function onChange(state: any, value: any) {
	return {
		...state,
		value,
		DropMenu: {
			...state.DropMenu,
			focusedIndex: null,
		},
	};
}

export function onSelect(state: any, selectedIndex: any) {
	const value = _.get(state.suggestions, selectedIndex);

	return onChange(
		{
			...state,
			value,
			DropMenu: {
				...DropMenu.onSelect(state.DropMenu, selectedIndex),
				selectedIndices: [],
			},
		},
		value
	);
}

export function onExpand(state: any) {
	return {
		...state,
		DropMenu: {
			...state.DropMenu,
			focusedIndex: null,
			isExpanded: !_.isEmpty(state.value) && !_.isEmpty(state.suggestions),
		},
	};
}

export { DropMenu };
