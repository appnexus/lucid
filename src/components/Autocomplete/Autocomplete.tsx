import React from 'react';
import PropTypes from 'react-peek/prop-types';
import _ from 'lodash';
import { createClass, StandardProps } from '../../util/component-types';
import { lucidClassNames } from '../../util/style-helpers';
import { buildHybridComponent } from '../../util/state-management';
import { partitionText } from '../../util/text-manipulation';
import * as reducers from './Autocomplete.reducers';
import * as KEYCODE from '../../constants/key-code';
import { DropMenuDumb as DropMenu, IDropMenuProps } from '../DropMenu/DropMenu';

const cx = lucidClassNames.bind('&-Autocomplete');

const { arrayOf, bool, func, object, shape, string } = PropTypes;

interface IAutocompleteProps extends StandardProps {
	isDisabled?: boolean;
	suggestions?: string[];
	value?: any;   
	onChange?: any;
	onSelect?: any;
	onExpand?: any;
	DropMenu?: IDropMenuProps;
}

const Autocomplete = createClass<IAutocompleteProps>({
	statics: {
		peek: {
			description: `
				A text input with suggested values displayed in an attached menu.
			`,
			categories: ['controls', 'text'],
			madeFrom: ['DropMenu'],
		},
	},

	displayName: 'Autocomplete',

	reducers: reducers as any, // TODO: typescript hack that should be removed 

	propTypes: {
		className: string`
			Appended to the component-specific class names set on the root elements.
		`,

		style: object`
			Styles that are passed through to root element.
		`,

		isDisabled: bool`
			Disables the Autocomplete from being clicked or focused.
		`,

		suggestions: arrayOf(string)`
			Array of suggested text input values shown in drop menu.
		`,

		value: string`
			Text value of the input.
		`,

		DropMenu: shape(DropMenu.propTypes)`
			Object of DropMenu props which are passed thru to the underlying DropMenu
			component.
		`,

		onChange: func`
			Called when the input value changes. Has the signature
			\`(value, {props, event}) => {}\` where value is a string.
		`,

		onSelect: func`
			Called when a suggstion is selected from the menu. Has the signature
			\`(optionIndex, {props, event}) => {}\` where optionIndex is a number.
		`,

		onExpand: func`
			Called when menu is expected to expand. Has the signature
			\`({props, event}) => {}\`.
		`,
	} as any, // TODO: typescript hack that should be removed

	getDefaultProps() {
		return {
			isDisabled: false,
			suggestions: [],
			value: '',
			onChange: _.noop,
			onSelect: _.noop,
			onExpand: _.noop,
			DropMenu: DropMenu.defaultProps,
		} as any; // TODO: typescript hack that should be removed
	},

	handleSelect(optionIndex: number, { event }: any) {
		const { suggestions, onChange, onSelect } = this.props;

		onChange(suggestions[optionIndex], { event, props: this.props });
		onSelect(optionIndex, { event, props: this.props });
	},

	handleInput(event: any) {
		const {
			onChange,
			onExpand,
			DropMenu: { onCollapse },
		} = this.props;

		onChange(event.target.value, { event, props: this.props });
		if (!_.isEmpty(event.target.value)) {
			onExpand({ event, props: this.props });
		} else {
			onCollapse();
		}
	},

	getInputValue() {
		return _.get(this, 'inputRef.value', this.props.value);
	},

	setInputValue(value: any) {
		if (this.inputRef) {
			this.inputRef.value = value;
		}
	},

	handleInputKeydown(event:  React.KeyboardEvent<HTMLInputElement>) {
		const {
			onExpand,
			DropMenu: { isExpanded, focusedIndex, onCollapse },
		} = this.props;

		const value = this.getInputValue();

		if (event.keyCode === KEYCODE.Tab && isExpanded && focusedIndex !== null) {
			this.handleSelect(focusedIndex, { event, props: this.props });
			event.preventDefault();
		}

		if (event.keyCode === KEYCODE.ArrowDown && !isExpanded) {
			event.stopPropagation();

			if (_.isEmpty(value)) {
				onExpand({ event, props: this.props });
			}
		}

		if (event.keyCode === KEYCODE.Escape) {
			event.stopPropagation();
			onCollapse(event);
		}

		if (event.keyCode === KEYCODE.Enter && focusedIndex === null) {
			event.stopPropagation();
			onCollapse(event);
		}
	},

	handleControlClick(event: React.MouseEvent<Element, MouseEvent>) {
		const {
			onExpand,
			DropMenu: { isExpanded, onCollapse },
		} = this.props;

		if (event.target === this.inputRef) {
			onExpand({ event, props: this.props });
		} else {
			if (isExpanded) {
				onCollapse(event);
			} else {
				onExpand({ event, props: this.props });
			}

			this.inputRef.focus();
		}
	},

	componentDidMount() {
		const { value } = this.props;
		this.inputRef.addEventListener('input', this.handleInput);
		this.setInputValue(value);
	},

	UNSAFE_componentWillReceiveProps(nextProps: any) { // TODO: typescript hack that should be removed
		const { value } = nextProps;
		if (value !== this.getInputValue()) {
			this.setInputValue(value);
		}
	},

	componentWillUnmount() {
		if (this.inputRef) {
			this.inputRef.removeEventListener('input', this.handleInput);
		}
	},

	render() {
		const {
			style,
			className,
			isDisabled,
			DropMenu: dropMenuProps,
			suggestions,
			...passThroughs
		} = this.props as any; // TODO: typescript hack that should be removed

		const { isExpanded } = dropMenuProps;

		const value = this.getInputValue();
		const valuePattern = new RegExp(_.escapeRegExp(value), 'i');

		return (
			<DropMenu
				{...dropMenuProps}
				isDisabled={isDisabled}
				selectedIndices={[]}
				className={cx('&', className)}
				onSelect={this.handleSelect}
				style={style}
			>
				<DropMenu.Control { ...{onClick: this.handleControlClick} /* TODO: typescript hack that should be removed */}> 
					<div
						className={cx('&-Control', {
							'&-Control-is-expanded': isExpanded,
							'&-Control-is-disabled': isDisabled,
						})}
					>
						<input
							{..._.omit(passThroughs, [
								'onChange',
								'onSelect',
								'onExpand',
								'value',
								'children',
							]) as any} // TODO: typescript hack that should be removed
							type='text'
							className={cx('&-Control-input')}
							ref={ref => (this.inputRef = ref)}
							onKeyDown={this.handleInputKeydown}
							disabled={isDisabled}
						/>
					</div>
				</DropMenu.Control>
				{value
					? _.map(suggestions, suggestion => (
							<DropMenu.Option key={'AutocompleteOption' + suggestion}>
								{(() => {
									const [pre, match, post] = partitionText(
										suggestion,
										valuePattern,
										value.length
									);
									const formattedSuggestion: any = [];
									if (pre) {
										formattedSuggestion.push(
											<span
												key={`AutocompleteOption-suggestion-pre-${suggestion}`}
												className={cx('&-Option-suggestion-pre')}
											>
												{pre}
											</span>
										);
									}
									if (match) {
										formattedSuggestion.push(
											<span
												key={`AutocompleteOption-suggestion-match-${suggestion}`}
												className={cx('&-Option-suggestion-match')}
											>
												{match}
											</span>
										);
									}
									if (post) {
										formattedSuggestion.push(
											<span
												key={`AutocompleteOption-suggestion-post-${suggestion}`}
												className={cx('&-Option-suggestion-post')}
											>
												{post}
											</span>
										);
									}
									return formattedSuggestion;
								})()}
							</DropMenu.Option>
					  ))
					: _.map(suggestions, suggestion => (
							<DropMenu.Option key={'AutocompleteOption' + suggestion}>
								{suggestion}
							</DropMenu.Option>
					  ))}
			</DropMenu>
		);
	},
});

export default buildHybridComponent(Autocomplete);
export { Autocomplete as AutocompleteDumb };
