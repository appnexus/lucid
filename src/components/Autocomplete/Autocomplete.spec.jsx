import React from 'react';
import { mount, shallow } from 'enzyme';
import assert from 'assert';
import sinon from 'sinon';
import { findTypes } from '../../util/component-types';
import _ from 'lodash';
import { common } from '../../util/generic-tests';
import Autocomplete from './Autocomplete';
import DropMenu from '../DropMenu/DropMenu';
import * as KEYCODE from '../../constants/key-code';

describe('Autocomplete', () => {
	common(Autocomplete);

	describe('render', () => {
		it('should render a DropMenu', () => {
			const wrapper = shallow(
				<Autocomplete />
			);

			assert.equal(wrapper.find('DropMenu').length, 1);
		});
	});

	describe('props', () => {
		describe('isDisabled', () => {
			it('should pass the `isDisabled` prop thru to the underlying DropMenu', () => {
				const wrapper = shallow(
					<Autocomplete isDisabled={true} />
				);

				const dropMenuWrapper = wrapper.find('DropMenu');

				assert.equal(dropMenuWrapper.prop('isDisabled'), true);
			});

			it('should apply the appropriate classNames to the control', () => {
				const wrapper = shallow(
					<Autocomplete isDisabled={true} />
				);

				const controlWrapper = wrapper.find('.lucid-Autocomplete-Control');

				assert(controlWrapper.hasClass('lucid-Autocomplete-Control-is-disabled'));
			});
		});

		describe('suggestions', () => {
			it('should create `DropMenu.Option`s for each suggestion and pass thru to underlying DropMenu', () => {
				const wrapper = shallow(
					<Autocomplete
						suggestions={[
							'Portland',
							'portal',
							'porridge',
							'potent',
							'please',
						]}
					/>
				);

				const options = _.map(findTypes(wrapper.find(DropMenu).props(), DropMenu.Option), 'props');

				assert.equal('Portland', options[0].children);
				assert.equal('portal', options[1].children);
				assert.equal('porridge', options[2].children);
				assert.equal('potent', options[3].children);
				assert.equal('please', options[4].children);
			});
		});

		describe('value', () => {
			let wrapper;
			let rootMountNode;

			beforeEach(() => {
				rootMountNode = document.createElement('div');
				document.body.appendChild(rootMountNode);
			});

			afterEach(() => {
				if (wrapper) {
					wrapper.unmount();
				}

				document.body.removeChild(rootMountNode);
			});

			it('should set the text value of the input', () => {
				wrapper = mount(
					<Autocomplete
						value='Portland'
					/>
				, { attachTo: rootMountNode });

				const inputDOMNode = document.querySelector('.lucid-Autocomplete-Control-input');

				assert.equal(inputDOMNode.value, 'Portland', 'input value must match prop value');
			});

			it('should change the text value of the input when the prop changes', () => {
				wrapper = mount(
					<Autocomplete
						value='Portland'
					/>
				, { attachTo: rootMountNode });

				wrapper.setProps({
					...wrapper.props(),
					value: 'Boston',
				})
				const inputDOMNode = document.querySelector('.lucid-Autocomplete-Control-input');

				assert.equal(inputDOMNode.value, 'Boston', 'input value must match bew prop value');
			});
		});



		describe('DropMenu', () => {
			it('should pass thru all DropMenu props to the underlying DropMenu', () => {
				const explicitDropMenuProps = {
					isExpanded: true,
					direction: 'up',
					focusedIndex: 2,
				};

				const wrapper = shallow(
					<Autocomplete DropMenu={explicitDropMenuProps} />
				);

				const dropMenuProps = wrapper.find('DropMenu').props();

				_.forEach(explicitDropMenuProps, (value, key) => {
					assert(_.isEqual(dropMenuProps[key], value));
				});
			});
		});

		describe('onChange', () => {
			let wrapper;
			let rootMountNode;

			beforeEach(() => {
				rootMountNode = document.createElement('div');
				document.body.appendChild(rootMountNode);
			});

			afterEach(() => {
				if (wrapper) {
					wrapper.unmount();
				}

				document.body.removeChild(rootMountNode);
			});


			it('should be called when a suggestion is selected from the menu', () => {
				const onChange = sinon.spy();

				wrapper = mount(
					<Autocomplete
						onChange={onChange}
						DropMenu={{ isExpanded: true }}
						suggestions={[
							'Portland',
							'portal',
							'porridge',
							'potent',
							'please',
						]}
						testProp='foo'
					/>
				);

				const menuDOMNode = document.querySelector('.lucid-ContextMenu-FlyOut');
				menuDOMNode.children[2].click();

				assert(onChange.called);
				const [textValue, { props, event }] = onChange.lastCall.args;
				assert.equal(textValue, 'porridge');
				assert(props);
				assert.equal(props.testProp, 'foo');
				assert(event);
			});

			it('should be called when user types into the text box', () => {
				const onChange = sinon.spy();

				wrapper = mount(
					<Autocomplete
						onChange={onChange}
						testProp='foo'
					/>
				, { attachTo: rootMountNode });

				const inputDOMNode = document.querySelector('.lucid-Autocomplete-Control-input');

				// set the input value and dispatch an `input` event
				inputDOMNode.value = 'aaa';
				const inputEvent = document.createEvent('Event');
				inputEvent.initEvent('input', true, true);
				inputDOMNode.dispatchEvent(inputEvent);

				assert(onChange.called, 'onChange must be called');
				const [textValue, { props, event }] = onChange.lastCall.args;
				assert.equal(textValue, 'aaa', 'value must match input');
				assert(props, 'props must be passed');
				assert.equal(props.testProp, 'foo', 'aditional props must be included');
				assert(event, 'event must be passed');
			});
		});

		describe('onSelect', () => {
			let wrapper;

			afterEach(() => {
				if (wrapper) {
					wrapper.unmount();
				}
			});


			it('should be called when a suggestion is selected from the menu', () => {
				const onSelect = sinon.spy();

				wrapper = mount(
					<Autocomplete
						onSelect={onSelect}
						DropMenu={{ isExpanded: true }}
						suggestions={[
							'Portland',
							'portal',
							'porridge',
							'potent',
							'please',
						]}
						testProp='foo'
					/>
				);

				const menuDOMNode = document.querySelector('.lucid-ContextMenu-FlyOut');
				menuDOMNode.children[2].click();

				assert(onSelect.called, 'onSelect must be called');
				const [optionIndex, { props, event }] = onSelect.lastCall.args;
				assert.equal(optionIndex, 2, 'optionIndex must be accurate');
				assert(props, 'props must be passed');
				assert.equal(props.testProp, 'foo', 'aditional props must be included');
				assert(event, 'event must be passed');
			});
		});

		describe('onExpand', () => {
			let wrapper;
			let rootMountNode;

			beforeEach(() => {
				rootMountNode = document.createElement('div');
				document.body.appendChild(rootMountNode);
			});

			afterEach(() => {
				if (wrapper) {
					wrapper.unmount();
				}

				document.body.removeChild(rootMountNode);
			});


			it('should be called when the input value changes to a non-empty value', () => {
				const onExpand = sinon.spy();

				wrapper = mount(
					<Autocomplete
						onExpand={onExpand}
						suggestions={[
							'Portland',
							'portal',
							'porridge',
							'potent',
							'please',
						]}
						testProp='foo'
					/>
				, { attachTo: rootMountNode });

				const inputDOMNode = document.querySelector('.lucid-Autocomplete-Control-input');

				// set the input value and dispatch an `input` event
				inputDOMNode.value = 'aaa';
				const inputEvent = document.createEvent('Event');
				inputEvent.initEvent('input', true, true);
				inputDOMNode.dispatchEvent(inputEvent);

				assert(onExpand.called, 'onExpand must be called');
				const [{ props, event }] = onExpand.lastCall.args;
				assert(props, 'props must be passed');
				assert.equal(props.testProp, 'foo', 'aditional props must be included');
				assert(event, 'event must be passed');
			});

			it('should be called when not expanded and down array key is pressed', () => {
				const onExpand = sinon.spy();

				const wrapper = shallow(
					<Autocomplete
						onExpand={onExpand}
						suggestions={[
							'Portland',
							'portal',
							'porridge',
							'potent',
							'please',
						]}
						testProp='foo'
						DropMenu={{
							isExpanded: false,
						}}
					/>
				);

				wrapper.find('.lucid-Autocomplete-Control-input').simulate('keydown', {
					keyCode: KEYCODE.ArrowDown,
					stopPropagation: _.noop,
				});

				assert(onExpand.called, 'onExpand must be called');
				const [{ props, event }] = onExpand.lastCall.args;
				assert(props, 'props must be passed');
				assert.equal(props.testProp, 'foo', 'aditional props must be included');
				assert(event, 'event must be passed');
			});

			it('should be called when the control input is clicked', () => {
				const onExpand = sinon.spy();

				wrapper = mount(
					<Autocomplete
						onExpand={onExpand}
						suggestions={[
							'Portland',
							'portal',
							'porridge',
							'potent',
							'please',
						]}
						testProp='foo'
					/>
				, { attachTo: rootMountNode });

				wrapper.find('.lucid-Autocomplete-Control-input').simulate('click');

				assert(onExpand.called, 'onExpand must be called');
				const [{ props, event }] = onExpand.lastCall.args;
				assert(props, 'props must be passed');
				assert.equal(props.testProp, 'foo', 'aditional props must be included');
				assert(event, 'event must be passed');
			});

			it('should be called when not expanded the non-input control is clicked', () => {
				const onExpand = sinon.spy();

				wrapper = mount(
					<Autocomplete
						onExpand={onExpand}
						suggestions={[
							'Portland',
							'portal',
							'porridge',
							'potent',
							'please',
						]}
						testProp='foo'
						DropMenu={{
							isExpanded: false,
						}}
					/>
				, { attachTo: rootMountNode });

				wrapper.find('.lucid-Autocomplete-Control').simulate('click');

				assert(onExpand.called, 'onExpand must be called');
				const [{ props, event }] = onExpand.lastCall.args;
				assert(props, 'props must be passed');
				assert.equal(props.testProp, 'foo', 'aditional props must be included');
				assert(event, 'event must be passed');
			});
		});
	});
});
