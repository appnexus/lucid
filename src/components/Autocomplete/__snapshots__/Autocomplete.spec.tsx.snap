// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Autocomplete [common] example testing should match snapshot(s) for Basic 1`] = `
<div
  class="lucid-DropMenu lucid-DropMenu-base lucid-Autocomplete"
>
  <span
    class="lucid-ContextMenu"
  >
    <div
      class="lucid-DropMenu-Control"
      tabindex="0"
    >
      <div
        class="lucid-Autocomplete-Control"
      >
        <input
          class="lucid-Autocomplete-Control-input"
          placeholder="Enter a word..."
          type="text"
        />
      </div>
    </div>
  </span>
</div>
`;

exports[`Autocomplete [common] example testing should match snapshot(s) for Disabled 1`] = `
<section
  style="min-height:100px"
>
  <div
    class="lucid-DropMenu lucid-DropMenu-base lucid-Autocomplete"
  >
    <span
      class="lucid-ContextMenu"
    >
      <div
        class="lucid-DropMenu-Control"
      >
        <div
          class="lucid-Autocomplete-Control lucid-Autocomplete-Control-is-disabled"
        >
          <input
            class="lucid-Autocomplete-Control-input"
            disabled=""
            placeholder="Enter a word..."
            type="text"
          />
        </div>
      </div>
    </span>
  </div>
</section>
`;

exports[`Autocomplete [common] example testing should match snapshot(s) for Interactive 1`] = `
<section>
  <section>
    Current Value: 
  </section>
  <div
    class="lucid-DropMenu lucid-DropMenu-base lucid-Autocomplete"
  >
    <span
      class="lucid-ContextMenu"
    >
      <div
        class="lucid-DropMenu-Control"
        tabindex="0"
      >
        <div
          class="lucid-Autocomplete-Control"
        >
          <input
            class="lucid-Autocomplete-Control-input"
            placeholder="Enter a word..."
            type="text"
          />
        </div>
      </div>
    </span>
  </div>
</section>
`;

exports[`Autocomplete [common] example testing should match snapshot(s) for Stateless 1`] = `
<div
  class="lucid-DropMenu lucid-DropMenu-base lucid-DropMenu-is-expanded lucid-DropMenu-direction-down lucid-Autocomplete"
>
  <span
    class="lucid-ContextMenu"
  >
    <div
      class="lucid-DropMenu-Control"
      tabindex="0"
    >
      <div
        class="lucid-Autocomplete-Control lucid-Autocomplete-Control-is-expanded"
      >
        <input
          class="lucid-Autocomplete-Control-input"
          placeholder="Enter a word..."
          type="text"
        />
      </div>
    </div>
  </span>
</div>
`;
